src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,551 @@
+    sas_uut_fad_object: FAD object from SAS UUT
+    sas_th_fad_object: a list of FAD objects from SAS Test Harness
+    pal_records: PAL records associated with a PPA protection area
+    pool: object of type Pool to distribute execution of a iapPointConstraint
+    function across multiple input values, distributing the input data across
+    processes
+   Returns:
+    ap_iap_ref: post IAP allowed interference in the format of 
+    {latitude : {longitude : [interference(mW/IAPBW), interference(mW/IAPBW)]}}
+  """
+  iap_threshold_ppa = Q_PPA_DBM_PER_RBW + \
+                    interf.linearToDb(interf.IAPBW_HZ / interf.PPA_RBW_HZ)
+  iap_margin_ppa = interf.dbToLinear(iap_threshold_ppa - MG_PPA_DB)
+
+  grant_objects = interf.getGrantObjectsFromFAD(sas_uut_fad_object, sas_th_fad_objects)
+  
+  # Get number of SAS 
+  num_sas = len(sas_th_fad_objects) + 1
+
+  logging.debug('$$$$ Getting GRID points for PPA Protection Area $$$$')
+
+  # Get Fine Grid Points for a PPA protection area
+  protection_points = utils.GridPolygon(protected_entity['zone']['features'][0]['geometry'], 2)
+
+  # Get the region type of the PPA protection area
+  ppa_region = protected_entity['ppaInfo']['ppaRegionType']
+
+  # Find the PAL records for the PAL_IDs defined in PPA records
+  ppa_pal_ids = protected_entity['ppaInfo']['palId']
+
+  for pal_record in pal_records:
+    if pal_record['palId'] == ppa_pal_ids[0]:
+      # Get the frequencies from the PAL records
+      ppa_freq_range = pal_record['channelAssignment']['primaryAssignment']
+      ppa_low_freq = ppa_freq_range['lowFrequency']
+      ppa_high_freq = ppa_freq_range['highFrequency']
+
+      # Get channels over which area incumbent needs partial/full protection
+      protection_channels = interf.getProtectedChannels(ppa_low_freq, ppa_high_freq)
+      
+      # Apply IAP for each protection constraint with a pool of parallel 
+      # processes.
+      logging.debug('$$$$ Calling PPA Protection $$$$')
+       
+      if not pool:
+        pool = Pool(processes=min(multiprocessing.cpu_count(), 
+                                   len(protection_points)))
+      asas_interference = interf.getInterferenceObject()
+      aggregate_interference = interf.getInterferenceObject()
+      iapPoint = partial(iapPointConstraint, channels=protection_channels, 
+                         low_freq=ppa_low_freq, high_freq=ppa_high_freq, 
+                         grant_objects=grant_objects, fss_info=None, 
+                         esc_antenna_info=None, region_type=ppa_region,
+                         threshold=iap_margin_ppa, 
+                         protection_ent_type=interf.ProtectedEntityType.PPA_AREA,
+                         asas_interference=asas_interference, 
+                         aggregate_interference=aggregate_interference) 
+
+      pool.map(iapPoint, protection_points)
+
+      break
+  
+  ap_iap_ref = calculateApIapRef(iap_margin_ppa, num_sas,
```

Comment:
I think we also need to eventually distribute the 1db margin across all SASes.

Reply:
As per   point 3, of section 8.4.3.3 in TS-0061 , the actual IAP protection level for   protected entity = Q - Mg. As per the current implementation this value is   considered for IAP calculations                Point 3 of section 8.4.3.3 also mentions, allowed headroom at SAS UUT.   Allowed headroom is distributed eqaully across all peer SASs.           I=10*log{(〖10〗^(Q/10)-〖10〗^((Q-Mg)/10))/N_SAS }     Allowed headroom is considered to authorize grants while IAP is in   progress. This value is not used for calculation of post IAP allowed   aggregate interference.          Please clarify, if this understanding is not correct.